### ISSUE TITLE
Enrich release notes with trigger details 

### ISSUE THREAD

**Issue by rfennell** [rfennell/AzurePipelines#664]
### Azure DevOps Extension you are using
Xplatform Release Notes

### Enhancement 
This is extracted from the issue #658 
I do really need to be able to access Pull Request data as well for auditing purposes. I totally get you can't take all suggestions for enriching the data, but that would be super helpful, and I'm sure others would find it useful. Just a suggestion :)

**Comment by rfennell** [comment]
@kingua I have moved this enhancement to it's own issue. makes it easier to track

**Comment by rfennell** [comment]
I have some code that gets a [PR title that triggered a build](https://github.com/rfennell/AzurePipelines/tree/master/Extensions/ArtifactDescription) that I use in my release pipelines. It never made it as far as being published as an extension. Maybe I can re purpose that, I will check into that

**Comment by kingua** [comment]
Great!  Thanks for considering this @rfennell.  If possible, having info regarding who submitted the PR, the code review ID, who conducted the code review, and who approved the PR would be very helpful.  That way, there is an audit trail of not just the commits, but of the governance trails as well.

It looks like getting a list of all associated PRs and then looping to get each PR's details based on that list might be sufficient to just return 'all' the relevant PR related data.

**Comment by rfennell** [comment]
**What was already available**

Turns out there was already a means to get a limited set of PR details from a build. You used the form

```
**Build Trigger PR Number**: ${buildDetails.triggerInfo['pr.number']}
```

or in handlebars form

```
**Build Trigger PR Number**: {{lookup buildDetails.triggerInfo 'pr.number'}} 
```
<hr>

**The improvements**

 That said I have improved the options. There is now a new `prDetails` object available to the template.

If you use the dump option
```
${JSON.stringify(prDetails)}    
```
You can see the fields available
```
{
    "repository": {
        "id": "bebd0ae2-405d-4c0a-b9c5-36ea94c1bf59",
        "name": "VSTSBuildTaskValidation",
        "url": "https://richardfennell.visualstudio.com/670b3a60-2021-47ab-a88b-d76ebd888a2f/_apis/git/repositories/bebd0ae2-405d-4c0a-b9c5-36ea94c1bf59",
        "project": {
            "id": "670b3a60-2021-47ab-a88b-d76ebd888a2f",
            "name": "GitHub",
            "description": "A container for GitHub CI/CD processes",
            "url": "https://richardfennell.visualstudio.com/_apis/projects/670b3a60-2021-47ab-a88b-d76ebd888a2f",
            "state": "wellFormed",
            "revision": 411511726,
            "visibility": 2,
            "lastUpdateTime": "2019-10-10T20:35:51.85Z"
        },
        "size": 9373557,
        "remoteUrl": "https://richardfennell.visualstudio.com/DefaultCollection/GitHub/_git/VSTSBuildTaskValidation",
        "sshUrl": "richardfennell@vs-ssh.visualstudio.com:v3/richardfennell/GitHub/VSTSBuildTaskValidation",
        "webUrl": "https://richardfennell.visualstudio.com/DefaultCollection/GitHub/_git/VSTSBuildTaskValidation"
    },
    "pullRequestId": 4,
    "codeReviewId": 4,
    "status": 1,
    "createdBy": {
        "displayName": "Richard Fennell (Work MSA)",
        "url": "https://spsprodeus24.vssps.visualstudio.com/Ac0efb61e-a937-42a0-9658-649757d55d46/_apis/Identities/b1fce0e9-fbf4-4202-bc09-a290def3e98b",
        "_links": {
            "avatar": {
                "href": "https://richardfennell.visualstudio.com/_apis/GraphProfile/MemberAvatars/aad.NzQzY2UyODUtN2Q0Ny03YjNkLTk0ZGUtN2Q0YjA1ZGE5NDdj"
            }
        },
        "id": "b1fce0e9-fbf4-4202-bc09-a290def3e98b",
        "uniqueName": "bm-richard.fennell@outlook.com",
        "imageUrl": "https://richardfennell.visualstudio.com/_api/_common/identityImage?id=b1fce0e9-fbf4-4202-bc09-a290def3e98b",
        "descriptor": "aad.NzQzY2UyODUtN2Q0Ny03YjNkLTk0ZGUtN2Q0YjA1ZGE5NDdj"
    },
    "creationDate": "2020-04-04T10:44:59.566Z",
    "title": "Added test.txt",
    "description": "Added test.txt",
    "sourceRefName": "refs/heads/branch2",
    "targetRefName": "refs/heads/master",
    "mergeStatus": 3,
    "isDraft": false,
    "mergeId": "f76a6556-8b4f-44eb-945a-9350124f067b",
    "lastMergeSourceCommit": {
        "commitId": "f43fa4de163c3ee0b4f17b72a659eac0d307deb8",
        "url": "https://richardfennell.visualstudio.com/670b3a60-2021-47ab-a88b-d76ebd888a2f/_apis/git/repositories/bebd0ae2-405d-4c0a-b9c5-36ea94c1bf59/commits/f43fa4de163c3ee0b4f17b72a659eac0d307deb8"
    },
    "lastMergeTargetCommit": {
        "commitId": "829ab2326201c7a5d439771eef5a57f58f94897d",
        "url": "https://richardfennell.visualstudio.com/670b3a60-2021-47ab-a88b-d76ebd888a2f/_apis/git/repositories/bebd0ae2-405d-4c0a-b9c5-36ea94c1bf59/commits/829ab2326201c7a5d439771eef5a57f58f94897d"
    },
    "lastMergeCommit": {
        "commitId": "53f393cae4ee3b901bb69858c4ee86cc8b466d6f",
        "author": {
            "name": "Richard Fennell (Work MSA)",
            "email": "bm-richard.fennell@outlook.com",
            "date": "2020-04-04T10:44:59.000Z"
        },
        "committer": {
            "name": "Richard Fennell (Work MSA)",
            "email": "bm-richard.fennell@outlook.com",
            "date": "2020-04-04T10:44:59.000Z"
        },
        "comment": "Merge pull request 4 from branch2 into master",
        "url": "https://richardfennell.visualstudio.com/670b3a60-2021-47ab-a88b-d76ebd888a2f/_apis/git/repositories/bebd0ae2-405d-4c0a-b9c5-36ea94c1bf59/commits/53f393cae4ee3b901bb69858c4ee86cc8b466d6f"
    },
    "reviewers": [],
    "url": "https://richardfennell.visualstudio.com/670b3a60-2021-47ab-a88b-d76ebd888a2f/_apis/git/repositories/bebd0ae2-405d-4c0a-b9c5-36ea94c1bf59/pullRequests/4",
    "supportsIterations": true,
    "artifactId": "vstfs:///Git/PullRequestId/670b3a60-2021-47ab-a88b-d76ebd888a2f%2fbebd0ae2-405d-4c0a-b9c5-36ea94c1bf59%2f4"
}
```
and in templates this is used 
```
**PR Title **: ${prDetails.title}
```
or in handlebars format.

```
**PR Details**: {{prDetails.title}}
```

This has been released a 2.33.3

Interested to here feedback from the real world as opposed to test harnesses

**Comment by kingua** [comment]
@rfennell, I see that the task as been updated to 2.33.3 in our Azure DevOps release pipeline.  When I use ${JSON.stringify(prDetails)} to dump all the fields, nothing is coming back, just an empty object {}.  Do you have any ideas?

**Comment by rfennell** [comment]
Remember, it will only work if the source is stored on an Azure DevOps repo

In the log you should see a message in the form

> The default artifact for the release was triggered by the PR 1234, getting details
 
If you don't it will not have even tried to get PR details

The logic to run this is in GenerateReleaseNotes.ts Ln 285

**Comment by kingua** [comment]
Hm...okay.  The source code is in an Azure DevOps repo.  I'm not seeing the call to get the PR details.

We are releasing on merge, after a PR has been approved and accepted, not on the PR submittal itself.  That might be the issue?

I'll look at the code and experiement.  Thanks.

**Comment by kingua** [comment]
Am I understanding correctly that the way the PR lookup is currently working depends on a build policy being set on a branch to build on PR in order to get this to work?  

If so, it it possible instead, to just pull the associated PR on a normal CI/CD on commit process?

**Comment by rfennell** [comment]
The issue is finding the PR ID, I am currently getting it from the primary artifacts build `currentBuild.triggerInfo["pr.number"]`

You can see the data you have available using an API call something like

``
https://dev.azure.com/<org>/<project>/_apis/build/builds/123
``
and checking the returned JSON for PR details.

So I guess what you ask is possible, assuming that this value is present, if it is not then not sure what can be done

**Comment by rfennell** [comment]
OK, that comes at the problem a different way. You are right that I had focused on the assumption that the pipeline would be triggered by a PR validation build. I will look at adding you suggestion too

**Comment by kingua** [comment]
@rfennell, thank you, sir!  I can't express how appreciative I am of you enhancing this already great task :)

**Comment by rfennell** [comment]
@kingua 2.34.7 now should add a list of all the PRs that share a `LastMergeCommit ID` with any commit associated with the release. I have checked that this works for Azure DevOps based PRs using

- Squas
- Rebase
- Merge
- Linear

Now an array `pullRequests` is passed into the template processor. This can be accessed using a [Handlebars based template](https://github.com/rfennell/AzurePipelines/blob/master/SampleTemplates/XplatGenerateReleaseNotes%20(Node%20based)/Version%202/release-handlebars-template.md). I have decided not to do the work to provide support the new PR array in the legacy templates - it seems a waste of effort given that handlebars is so much more flexible. I think this will be the first step to deprecating the legacy templating system

A template block would look something like

```
### Associated Pull Requests ({{pullRequests.length}})
{{#forEach pullRequests}}
{{#if isFirst}}### Associated Pull Requests (only shown if  PR) {{/if}}
*  **PR {{this.id}}**  {{this.title}}
{{/forEach}}
```

Let me know how it goes

**Comment by kingua** [comment]
@rfennell, this was working perfectly yesterday, but it's stopped finding any PRs as of today.  I just noticed you submitted a PR to make some changes to the PR matching logic.  It looks like the latest change might have broken something ;)

**Comment by rfennell** [comment]
The logic change stops looking for PRs unless it is a tfsgit repo. And also different logic to match commit ids to PRs as some users are getting errors.

The new release passed my tests, so could you upload a log with system.debug=true so I can see what it is doing

**Comment by kingua** [comment]
Please see attached.
[Log.txt](https://github.com/rfennell/AzurePipelines/files/4489392/Log.txt)

It is failing to find the associated PR(s) where it was yesterday.  The repos are entirely within Azure DevOps Git.

**Comment by rfennell** [comment]
On first look at the log the call to find all completed PRs is not returning any PRs to attempt to match on. 

The code to find the list of all PRs is scoped to the current team project, could this be the issue

**Comment by kingua** [comment]
Yes!  That's it!  We have about 40 applications in separate projects and then a single project for our DevOps pipelines, so we can 'try' to enforce some some best practices, commonality, re-use, etc.  Is it possible to go back to allowing cross project search?

**Comment by rfennell** [comment]
I don't want to make cross project the default due to performance, but I can make it optional with a new parameter. I will do it this morning

**Comment by rfennell** [comment]
The change has been made and published as 2.37.1.

The default behaviour is unchanged, but if you set the advanced option searchCrossProjectForPRs to true it will do a cross team project search

I am going to close this issue now as I think the enhancement is done. If there are problems log a new issue

**Comment by kingua** [comment]
@rfennell, this seems to be functioning perfectly!  Thanks again!  Cheers!
